idx/memory/memory.go

```diff
@@ -130,6 +147,93 @@ func (m *MemoryIdx) Update(entry idx.Archive) {
 	m.Unlock()
 }
 
+// reads the tags of a given metric definition and creates the
+// corresponding tag index entries to refer to it.
+// assumes a lock is already held.
+func (m *MemoryIdx) indexTags(def *schema.MetricDefinition) {
+	tags, ok := m.Tags[def.OrgId]
+	if !ok {
+		tags = make(TagIndex)
+		m.Tags[def.OrgId] = tags
+	}
+	for _, tag := range def.Tags {
+		tagSplits := strings.SplitN(tag, "=", 2)
+		if len(tagSplits) < 2 {
+			// should never happen because every tag in the index
+			// must have a valid format
+			InvalidTagInIndex.Inc()
+			continue
+		}
+
+		tagName := tagSplits[0]
+		tagValue := tagSplits[1]
+
+		if _, ok = tags[tagName]; !ok {
+			tags[tagName] = make(TagValue)
+		}
+
+		if _, ok = tags[tagName][tagValue]; !ok {
+			tags[tagName][tagValue] = make(TagIDs)
+		}
+
+		id, err := idx.NewMetricIDFromString(def.Id)
+		if err != nil {
+			// should never happen because all IDs in the index must have
+			// a valid format
+			CorruptIndex.Inc()
+			continue
+		}
+		tags[tagName][tagValue][id] = struct{}{}
+	}
+}
+
+// takes a given metric definition and removes all references
+// to it from the tag index.
+// assumes a lock is already held.
+func (m *MemoryIdx) deindexTags(def *schema.MetricDefinition) {
+	tags, ok := m.Tags[def.OrgId]
+	if !ok {
+		return
+	}
+
+	for _, tag := range def.Tags {
+		tagSplits := strings.SplitN(tag, "=", 2)
+		if len(tagSplits) < 2 {
+			// should never happen because every tag in the index
+			// must have a valid format
```

Comment:
but we don't validate incoming MetricData's ? seems like the validation methods in the schema repo should be updated.
note that we currently allow single-word tags like `prod` which are not `k=v` pairs, but i'm okay with enforcing `k=v` format.

Reply:
Actually the method `indexTags()` does kind of verify the format, at least it won't add tags to the tag index if the format is invalid (although the invalid tags would still end up as part of the metric definition). 
Yeah, we should definitely update the validation methods in the schema repo, will submit a PR.